theano/compile/debugmode.py

```diff
@@ -2311,6 +2312,24 @@ def __init__(self, inputs, outputs, mode,
                               "of", len(li), "events was stable.",
                               file=sys.stderr)
         self.fgraph = fgraph
+        destroy_handler_added = False
```

Comment:
In fact, I think all of that block should be skipped if `cycle_detection=='fast'`, as trying to insert `_output_guard` that way does not actually insert a new cycle, and is not detected as a cycle by the 'fast' algo.
To be more precise: the old cycle detection prevents working inplace on outputs that existed _before_ the optimization was applied. The "fast" one considers the outputs _after_ the substitution, so there's no issue.